## Snakefile


    ## module load snakemake/5.2.2 (also loads anaconda and python3)
    ## snakemake -np TARGET


## globals
anno = '/scratch/chd5n/aneuploidy/raw-data/annotations/'


## rules
rule write_annotations:
    input:
        'gdc_write_anno.py'
    output:
        anno + 'pheno.tsv',
        anno + 'manifest.tsv'
    shell:
        'python {input}'

rule parse_info:
    input:
        'gdc_parse_info.py'
    output:
        anno + 'coad-read.file_info',
        anno + 'coad-read.sub_errors'
    shell:
        'python {input}'

rule download:
    input:
    output:
    shell:


# ## example format
# rule write_annotation:
#     input:
#         "data/genome.fa",
#         "data/samples/A.fastq"
#     output:
#         "mapped_reads/A.bam"
#     shell:
#         "bwa mem {input} | samtools view -Sb - > {output}"
